### PR TITLE
Enable flexible scaling on lso cluster

### PR DESCRIPTION
**Pull request by OdedViner**
Signed-off-by: Oded Viner <oviner@redhat.com>

### CODE REVIEW

File: ocs_ci/deployment/deployment.py [hunk 1]
```diff
@@ -491,6 +491,18 @@ def deploy_ocs_via_operator(self, image=None):
...
+        if config.ENV_DATA.get("enable_flexible_scaling") and config.DEPLOYMENT.get(
+            "local_storage"
+        ):
+            ocs_version = float(config.ENV_DATA["ocs_version"])
+            if ocs_version < 4.7:
+                error_message = "Flexible scaling can be enabled only on OCS >= 4.7!"
+                logger.error(error_message)
+                raise UnsupportedFeatureError(error_message)
+            else:
+                cluster_data["spec"]["flexibleScaling"] = True
```
Comment (petr-balogh) [inline ocs_ci/deployment/deployment.py:504]: you can avoid else block and indent one left I think

**Review by clacroix12 — APPROVED** [review]

**Comment by OdedViner** [comment]
```
$ oc get pv | grep local
local-pv-1994225b                          100Gi      RWO            Delete           Available                                                                    localblock                             84m
local-pv-1a9f4452                          100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-1-data-072kh8                    localblock                             84m
local-pv-6c9c36a                           100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-0-data-0pjj68                    localblock                             84m
local-pv-7e48d6e1                          100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-2-data-06lp9t                    localblock                             84m
************************

$ oc get pvc -A | grep -i local
openshift-storage          ocs-deviceset-0-data-0pjj68                 Bound    local-pv-6c9c36a                           100Gi      RWO            localblock                    81m
openshift-storage          ocs-deviceset-1-data-072kh8                 Bound    local-pv-1a9f4452                          100Gi      RWO            localblock                    81m
openshift-storage          ocs-deviceset-2-data-06lp9t                 Bound    local-pv-7e48d6e1                          100Gi      RWO            localblock                    81m
************************

$ oc get storagecluster -n openshift-storage -o yaml | grep flex
          f:flexibleScaling: {}
    flexibleScaling: true
************************

$ oc get nodes --show-labels | grep cluster.ocs.openshift.io/openshift-storage= |cut -d' ' -f1
compute-0
compute-1
compute-2

$ oc get nodes
NAME              STATUS   ROLES    AGE    VERSION
compute-0         Ready    worker   132m   v1.20.0+ba45583
compute-1         Ready    worker   132m   v1.20.0+ba45583
compute-2         Ready    worker   132m   v1.20.0+ba45583
compute-3         Ready    worker   132m   v1.20.0+ba45583
control-plane-0   Ready    master   139m   v1.20.0+ba45583
control-plane-1   Ready    master   139m   v1.20.0+ba45583
control-plane-2   Ready    master   139m   v1.20.0+ba45583


***************************
$ oc get cephblockpool -n openshift-storage -o yaml | grep failureDomain
          f:failureDomain: {}
    failureDomain: host

***************************
All nodes [0,1,2,3] attached disk (100G HDD) (check via vcenter)
***************************

Manually:
Edit storagecluster yaml file, Change count field:
$ oc get storagecluster -n openshift-storage -o yaml | grep -i count
      count: 1

Change Count=4

$ oc get pods -n openshift-storage  | grep osd
rook-ceph-osd-0-6bd64b94d7-ggrbf                                  2/2     Running     0          100m
rook-ceph-osd-1-6b6dc4db88-fbzc5                                  2/2     Running     0          100m
rook-ceph-osd-2-5c646fbffc-pxrch                                  2/2     Running     0          100m
rook-ceph-osd-prepare-ocs-deviceset-0-data-0pjj68-4l2hp           0/1     Completed   0          100m
rook-ceph-osd-prepare-ocs-deviceset-0-data-1fsttf-rcrqf           0/1     Pending     0          3m4s
rook-ceph-osd-prepare-ocs-deviceset-0-data-2cgxc4-cc2sw           0/1     Pending     0          3m3s
rook-ceph-osd-prepare-ocs-deviceset-0-data-37wzg5-s4plh           0/1     Pending     0          3m3s
rook-ceph-osd-prepare-ocs-deviceset-1-data-072kh8-pwhkq           0/1     Completed   0          100m
rook-ceph-osd-prepare-ocs-deviceset-1-data-1jsqk8-tq8t4           0/1     Pending     0          3m2s
rook-ceph-osd-prepare-ocs-deviceset-1-data-2mcnf8-vzx7w           0/1     Pending     0          3m2s
rook-ceph-osd-prepare-ocs-deviceset-1-data-3f7tmb-85bzc           0/1     Pending     0          3m1s
rook-ceph-osd-prepare-ocs-deviceset-2-data-06lp9t-nppgs           0/1     Completed   0          100m
rook-ceph-osd-prepare-ocs-deviceset-2-data-1jbjhb-bcdgg           0/1     Pending     0          3m
rook-ceph-osd-prepare-ocs-deviceset-2-data-2ftgpk-xdxwx           0/1     Pending     0          3m
rook-ceph-osd-prepare-ocs-deviceset-2-data-3sdtjf-7hshz           0/1     Pending     0          3m
```

**Comment by OdedViner** [comment]
```
$ oc get pvc -A | grep local
openshift-storage          ocs-deviceset-0-data-0pjj68                 Bound     local-pv-6c9c36a                           100Gi      RWO            localblock                    104m
openshift-storage          ocs-deviceset-0-data-1fsttf                 Pending                                                                        localblock                    6m41s
openshift-storage          ocs-deviceset-0-data-2cgxc4                 Pending                                                                        localblock                    6m41s
openshift-storage          ocs-deviceset-0-data-37wzg5                 Pending                                                                        localblock                    6m41s
openshift-storage          ocs-deviceset-1-data-072kh8                 Bound     local-pv-1a9f4452                          100Gi      RWO            localblock                    104m
openshift-storage          ocs-deviceset-1-data-1jsqk8                 Pending                                                                        localblock                    6m40s
openshift-storage          ocs-deviceset-1-data-2mcnf8                 Pending                                                                        localblock                    6m40s
openshift-storage          ocs-deviceset-1-data-3f7tmb                 Pending                                                                        localblock                    6m40s
openshift-storage          ocs-deviceset-2-data-06lp9t                 Bound     local-pv-7e48d6e1                          100Gi      RWO            localblock                    104m
openshift-storage          ocs-deviceset-2-data-1jbjhb                 Pending                                                                        localblock                    6m40s
openshift-storage          ocs-deviceset-2-data-2ftgpk                 Pending                                                                        localblock                    6m40s
openshift-storage          ocs-deviceset-2-data-3sdtjf                 Pending                                                                        localblock                    6m39s
[odedviner@localhost auth]$ oc get pv | grep local
local-pv-1994225b                          100Gi      RWO            Delete           Available                                                                    localblock                             106m
local-pv-1a9f4452                          100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-1-data-072kh8                    localblock                             106m
local-pv-6c9c36a                           100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-0-data-0pjj68                    localblock                             106m
local-pv-7e48d6e1                          100Gi      RWO            Delete           Bound       openshift-storage/ocs-deviceset-2-data-06lp9t                    localblock                             106m

```

File: ocs_ci/deployment/deployment.py [hunk 2]
```diff
@@ -491,6 +491,17 @@ def deploy_ocs_via_operator(self, image=None):
 
         cluster_data["metadata"]["name"] = config.ENV_DATA["storage_cluster_name"]
 
+        # Flexible scaling is available from version 4.7
+        if config.ENV_DATA.get("enable_flexible_scaling") and config.DEPLOYMENT.get(
+            "local_storage"
+        ):
+            ocs_version = float(config.ENV_DATA["ocs_version"])
```
Comment (petr-balogh) [inline ocs_ci/deployment/deployment.py:498]: If this enable_flexible_scaling should be really enabled by default for each LSO cluster on version >= 4.7 we should change condition to something like this:
```suggestion
        ocs_version = float(config.ENV_DATA["ocs_version"])
        if config.ENV_DATA.get("enable_flexible_scaling") or (
            config.DEPLOYMENT.get("local_storage") and ocs_version >= 4.7)
        ):
```

File: ocs_ci/deployment/deployment.py [hunk 3]
```diff
@@ -491,6 +491,17 @@ def deploy_ocs_via_operator(self, image=None):
...
+            if ocs_version < 4.7:
+                error_message = "Flexible scaling can be enabled only on OCS >= 4.7!"
+                logger.error(error_message)
+                raise UnsupportedFeatureError(error_message)
+            cluster_data["spec"]["flexibleScaling"] = True
```
[the remark this answers is not in the document]
Comment (petr-balogh) [reply ocs_ci/deployment/deployment.py:503]: Also add to the comment this BZ:
https://bugzilla.redhat.com/show_bug.cgi?id=1921023#c4
So we will know why we are changing this.

File: ocs_ci/deployment/deployment.py [hunk 4]
```diff
@@ -491,6 +491,17 @@ def deploy_ocs_via_operator(self, image=None):
...
+
```
Comment (nbalacha) [inline ocs_ci/deployment/deployment.py:504]: The storagecluster.yaml will also need to be modified to set the count =3 and replica=1

Comment (petr-balogh) [reply ocs_ci/deployment/deployment.py:504]: @nbalacha I commented this out above:  https://github.com/red-hat-storage/ocs-ci/pull/3826/files#r572789079 so I am resolving this comment as Oded will handle this in my older comment. Thanks

File: conf/deployment/vsphere/upi_1az_rhcos_vsan_lso_vmdk_3m_4w.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+---
```
[the remark this answers is not in the document]
Comment (petr-balogh) [reply conf/deployment/vsphere/upi_1az_rhcos_vsan_lso_vmdk_3m_4w.yaml:1]: For production jobs we would like to have production configuration file under deployment folder where we have everything in one config file. So I am for this configuration file as if we need production job we will use this one.

**Review by ebenahar — APPROVED** [review]

**Review by yitzhak12 — APPROVED** [review]

File: ocs_ci/ocs/resources/storage_cluster.py [hunk 6]
```diff
@@ -340,23 +341,16 @@ def ocs_install_verification(
...
-        ]
-        for crush_rule in crush_rules:
-            assert [
-                item for item in crush_rule["steps"] if item.get("type") == "zone"
-            ], f"{crush_rule['rule_name']} is not with type as zone"
-        log.info("Verified - pool crush rule is with type: zone")
+        pool_crush_verification(pool_type="zone")
+    elif (
+        config.DEPLOYMENT.get("local_storage")
+        and ocs_version >= 4.7
```
Comment (yitzhak12) [inline ocs_ci/ocs/resources/storage_cluster.py:349]: The var `ocs_version` is currently the 'str' type, and we need to convert it to float. maybe we can replace the line above with this line instead: 
`and float(ocs_version) >= 4.7`

**Pull request closed without merging**